### PR TITLE
[KOGITO-7293] - CNCF Serverless Workflow Specification Support

### PR DESCRIPTION
**Pull request by ricardozanini**
See: https://issues.redhat.com/browse/KOGITO-7293

- Added a new extension to support emojis 🎉 
- Added Mac garbage files to gitignore
- Introduced a few more variables

Must be merged after:

- #8 

Won't block docs review.

### CODE REVIEW

**Review by spolti — CHANGES_REQUESTED** [review]

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,4268 @@
+{
```
[the remark this answers is not in the document]
Comment (ricardozanini) [reply package-lock.json:1]: It's: 

```

NAME
       package-lock.json - A manifestation of the manifest
 
   Description
       package-lock.json is automatically generated for any operations where npm modifies either the node_modules tree, or package.json. It describes the exact tree that was generated, such that
       subsequent installs are able to generate identical trees, regardless of intermediate dependency updates.
 
       This file is intended to be committed into source repositories, and serves various purposes:
 
       • Describe a single representation of a dependency tree such that teammates, deployments, and continuous integration are guaranteed to install exactly the same dependencies.
 
       • Provide a facility for users to "time-travel" to previous states of node_modules without having to commit the directory itself.
 
       • Facilitate greater visibility of tree changes through readable source control diffs.
 
       • Optimize the installation process by allowing npm to skip repeated metadata resolutions for previously-installed packages.
 
       • As of npm v7, lockfiles include enough information to gain a complete picture of the package tree, reducing the need to read package.json files, and allowing for significant performance
         improvements.

```

Source: npm help package-lock.json.

File: antora-playbook.yml [hunk 2]
```diff
@@ -15,8 +15,15 @@ ui:
...
 asciidoc:
+  extensions:
+  - asciidoctor-emoji
   attributes:
     linkattrs: ''
     toc: ~
     quarkus_version: '2.9.1.Final'
-    required_maven_version: '3.8.1+'
\ No newline at end of file
+    required_maven_version: '3.8.1+'
```
[the remark this answers is not in the document]
Comment (ricardozanini) [reply antora-playbook.yml:24]: Just added there after rebasing we can merge altogether and won't mess up the order.

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 3]
```diff
@@ -1 +1,368 @@
...
+Kogito supports workflow secrets according to the {spec_doc_url}#workflow-secrets[specification definition].
+
+IMPORTANT: Secrets are tied to the {quarkus-config-guide-url}[Quarkus Configuration] module. Meaning that any form of configuration that Quarkus supports, it is also available in Kogito.
+
+See the guide "xref:core/accessing-workflow-metainformation-in-runtime.adoc[Accessing Workflow Metainformation]" for more details.
+
+== What's Next?
+
+- xref:getting-started/cncf-serverless-workflow-specification-support.adoc[Creating your first Serverless Workflow service]
+- xref:getting-started/getting-familiar-with-our-tooling.adoc[Getting familiar with our tooling].
```
[the remark this answers is not in the document]
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:368]: Nops, it will break my build. Once we merge yours I will add it.

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 4]
```diff
@@ -1 +1,368 @@
...
+
+Currently, Kogito implements version {spec_doc_url}[{spec_version}] of the specification. The table below classifies the current implementation status for each specification feature. {spec_doc_url}#Retry-Definition[Retries] are the only feature that Kogito does not support yet.
+
+.Current implementation status of the specification features
+[cols="1,1,1"]
+|===
+| Feature | Status | Spec Reference
+
+| <<states>>
+| emoji:full_moon[]
```
Comment (ibek) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:24]: It would be good to set a tooltip/footnote what the emojis mean? https://docs.asciidoctor.org/asciidoc/latest/macros/icon-macro/#icon-macro-attributes-shared

full moon = fully implemented and compliant with the spec
etc...

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 5]
```diff
@@ -1 +1,368 @@
...
+:keywords: kogito, workflow, serverless, cncf, specification
+// links
+:quarkus_config_guide_url: https://quarkus.io/guides/config-reference
+
+
+In this guide, you can learn the details about the Kogito implementation of the {spec_website_url}[CNCF Serverless Workflow] specification.
+
+== Current Status
+
+Currently, Kogito implements version {spec_doc_url}[{spec_version}] of the specification. The table below classifies the current implementation status for each specification feature. {spec_doc_url}#Retry-Definition[Retries] are the only feature that Kogito does not support yet.
```
Comment (ibek) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:16]: It says that Retries are the only feature that Kogito does not support yet. However, workflow timeouts share the same emoji. How is it? Should we then change the emoji of workflow timeout to some experimental one? I believe it theoretically works with job service, right?

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 6]
```diff
@@ -1 +1,368 @@
...
+| emoji:full_moon[] 
+| {spec_doc_url}#Callback-State[Callback State]
+|===
+
+[#functions]
+=== Workflow Model - Functions
+
+The table below illustrates the current status of the functions Kogito supports. 
+
+For additional functions, the specification supports the `custom` type. Therefore, Kogito added support for `sysout` and `java` tyes. See <<func_sysout>> and <<func_java>> sections for more information.
```
Comment (ibek) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:118]: Typo in "tyes" => "types"

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 7]
```diff
@@ -1 +1,368 @@
...
+
+| graphql
+| emoji:construction[]
+| {spec_doc_url}#using-functions-for-graphql-service-invocations[Using Functions for GraphQL Service Invocations]
+
+| odata
+| emoji:construction[]
+| {spec_doc_url}#using-functions-for-odata-service-invocations[Using Functions for OData Service Invocations]
+
+| custom
```
Comment (ibek) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:149]: Should we discourage from using custom types in the guide?

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 8]
```diff
@@ -1 +1,368 @@
...
+| emoji:full_moon[]
+| {spec_doc_url}#defining-custom-function-types[Defining custom function types]. Kogito supports `java` and `sysout` custom functions.
+|===
+
+[#func_sysout]
```
Comment (ibek) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:154]: This is a guide about the spec. In my opinion the details of implementing custom functions do not belong here.

**Review by ibek — CHANGES_REQUESTED** [review]

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 9]
```diff
@@ -1 +1,368 @@
...
+
+| Operation
+| emoji:full_moon[] 
+| {spec_doc_url}#Operation-State[Operation State]
+
+| Switch
+| emoji:full_moon[] 
+| {spec_doc_url}#Switch-State[Switch State]
+
+| Delay
```
Comment (ibek) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:92]: Is it Delay state or Sleep state? I often see them interchanged. What should we be using? The link doesn't work because it should be "https://github.com/serverlessworkflow/specification/blob/0.8.x/specification.md#sleep-state"

Comment (ibek) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:92]: Is it really fully implemented? Someone commented in the constructs document that it also requires JobService so not sure if we should consider it done.

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 10]
```diff
@@ -1 +1,368 @@
...
+| emoji:full_moon[] 
+| {spec_doc_url}#Delay-State[Delay State]
+
+| Parallel
```
Comment (ibek) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:96]: Should we warn how Parallel state behaves in a single thread with maybe a partial implementation emote?

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 11]
```diff
@@ -1 +1,368 @@
...
+
+| custom
+| emoji:full_moon[]
+| {spec_doc_url}#defining-custom-function-types[Defining custom function types]. Kogito supports `java` and `sysout` custom functions.
+|===
+
+[#func_sysout]
+==== Sysout Custom Function
+
+You can use this `custom` function type for debugging reasons or quick demonstrations:
```
Comment (spolti) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:157]: debugging reasons of?

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc (hunk 5 above)
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:16]: Timeouts are not implemented yet, that's why it shares the same emoji. I'll change the text.

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 12]
```diff
@@ -1 +1,368 @@
...
+      "type": "custom", <2>
+      "operation": "service:java:com.acme.MyInterfaceOrClass::myMethod" <3>
+    }
+  ]
+}
+----
+
+1. The function name
+2. The `custom` function type
+3. The Kogito custom operation definition. `service` is the reserved operation keyword followed by the other keyword `java`. The next portion is the FQDN of the interface or implementation class, followed by the method name.
```
Comment (spolti) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:222]: Maybe it worth to describe what for FQDN stands for on the first time it appears.

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 13]
```diff
@@ -1 +1,368 @@
...
+    {
+      "name": "myState",
+      "type": "operation",
+      "actions": [
+        {
+          "name": "callJavaFunctionAction",
+          "functionRef": {
+            "refName": "myFunction",
+            "arguments": {
+              "data": "${ .my.expression.to.data }"
```
Comment (spolti) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:284]: maybe we can provide a working example for this use case?

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 14]
```diff
@@ -1 +1,368 @@
...
+Kogito supports the capabilities of the event according to the {spec_doc_url}#Event-Definition[specification defition of an event]. However, event `correlation` has limited support as per specification definition.
+
+For more details about the event correlation feature that Kogito supports, see the guide xref:eventing/event-correlation-with-workflows.adoc[Event Correlation with Workflows].
+
+[#workflow_data]
+=== Workflow Data
+
+Kogito has the full implementation of the data manipulation feature via either `jq` (default) or `jsonpath`.
+
+JSON Path can have some limitations in specific scenarios, mainly when the use case requires complex data manipulation. See the {spec_doc_url}#workflow-data[Workflow Data] chapter in the specification.
```
Comment (spolti) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:306]:
```suggestion
`JSON Path` can have some limitations in specific scenarios, mainly when the use case requires complex data manipulation. See the {spec_doc_url}#workflow-data[Workflow Data] chapter in the specification.
```

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 15]
```diff
@@ -1 +1,368 @@
...
+
+Kogito also supports xref:core/state-data-filtering[data filtering]. For more information, check the section {spec_doc_url}#state-data-filters[State Data Filters] in the specification documentation.
+
+[#expressions]
+=== Expressions
+
+Kogito supports either `jq` or `jsonpath` to define workflow expressions. As the {spec_doc_url}#workflow-expressions[specification defines], `jq` is the default expression language. If you want to use `jsonpath` instead, set the attribute `expressionLang` in the workflow definition:
```
Comment (spolti) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:313]: maybe rephrase as it is stated on line 304 as well?

**Review by spolti — CHANGES_REQUESTED** [review]
nice work, just a few more comments :)

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc (hunk 8 above)
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:154]: I thought about that, but since it's too short, I was afraid to break it into a separate guide. I can do that, I was just waiting for more opinions. :)

Please guys, +1 if we break this into another guide or -1 if not.

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc (hunk 9 above)
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:92]: Oh yeah, it was a mistake. I'll fix it.

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc (hunk 11 above)
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:157]: No, I think this sentence is correct

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc (hunk 13 above)
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:284]: I'll once I break this part into a new guide as requested by @ibek

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc (hunk 15 above)
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:313]: I think the other sentence will be in the new guide

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 16]
```diff
@@ -1 +1,368 @@
...
+      ]
+    }
+  ]
+}
+----
+
+[#events]
+=== Workflow Model - Events
+
+Kogito supports the capabilities of the event according to the {spec_doc_url}#Event-Definition[specification defition of an event]. However, event `correlation` has limited support as per specification definition.
```
Comment (krisv) [inline serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:297]: typo, definition

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc (hunk 4 above)
Comment (krisv) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:24]: +1

**Review by krisv — CHANGES_REQUESTED** [review]

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc (hunk 4 above)
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:24]: We have a legend in the original README: https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-serverless-workflow#current-status

I'll do that instead since emojis don't have tooltip. I looked for that yesterday.

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc (hunk 7 above)
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:149]: I don't think so.

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc [hunk 17]
```diff
@@ -1 +1,368 @@
...
+  ]
+}
+----
+
+1. The function name
+2. The `custom` function type
+3. The Kogito custom operation definition. `service` is the reserved operation keyword followed by the other keyword `java`. The next portion is the FQDN of the interface or implementation class, followed by the method name.
+
+Your method's interface *must* receive a https://github.com/FasterXML/jackson[Jackson's] `JsonNode` object and return either `void` or another `JsonNode`. For example:
+
```
[the remark this answers is not in the document]
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:225]: It will follow whatever you define in the next states. But yeah, the data returned here will be the next source of truth. I will highlight that.

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc (hunk 13 above)
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:284]: The example is already described in the guide.

**Comment by ricardozanini** [comment]
@krisv @ibek @spolti can you please do another look?

File: serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+      "type": "custom", <2>
+      "operation": "service:java:com.acme.MyInterfaceOrClass::myMethod" <3>
+    }
+  ]
+}
+----
+
+1. The function name
+2. The `custom` function type
+3. The Kogito custom operation definition. `service` is the reserved operation keyword followed by the other keyword `java`. The next portion is the FQDN (Fully Qualified Domain Name) of the interface or implementation class, followed by the method name.
```
Comment (krisv) [inline serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:85]: Shouldn't this be FQCN (class name)?

File: serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+        }
+      ]
+    }
+  ]
+}
+----
+
+The `JsonNode` that the function returns will be the new payload from this point of the workflow execution.
+
+WARNING: Refrain from using only Java functions because it can turn the workflow into a monolithic runtime application.
```
Comment (krisv) [inline serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:159]: Not sure it's always a bad idea, I guess a Java function that helps with handling some data doesn't seem too bad.  Just like scripts in BPMN2, they shouldn't however do more than that, so they really shouldn't invoke external services for example.

**Review by krisv — APPROVED** [review]

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc (hunk 14 above)
Comment (hmanwani-rh) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:186]: +1, JSON path doesn't seem anything related to code. However, I'd suggest not capitalizing on the path. I'd change it when I'll add my tweaks :)

File: serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc (hunk 19 above)
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:159]: @krisv maybe we can rephrase this block emphasizing that relying ONLY on java functions might not be a good idea?

**Review by spolti — APPROVED** [review]

File: serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc (hunk 14 above)
Comment (ricardozanini) [reply serverlessworkflow/modules/ROOT/pages/getting-started/cncf-serverless-workflow-specification-support.adoc:186]: @hmanwani-rh the right name is "JSONPath", I'll use that instead: https://goessner.net/articles/JsonPath/

**Review by ibek — APPROVED** [review]

**Review by cristianonicolai — APPROVED** [review]

**Review by akumar074 — APPROVED** [review]

**Review by amadhusu — APPROVED** [review]

**Comment by ricardozanini** [comment]
Rebased with @spolti changes.

File: serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc [hunk 20]
```diff
@@ -154,13 +165,16 @@ Alternatively, you can only pass the necessary data:
 }
 ----
 
-The `JsonNode` that the function returns will be the new payload from this point of the workflow execution.
+In the previous example, the `JsonNode` that the `java` function returns is the new payload in the Serverless Workflow execution.
 
-WARNING: Refrain from using **only** Java functions or use them to call external services. Try to use the xref:service-orchestration/orchestration-of-opnapi-based-services.adoc[the services orchestration features instead].
+[WARNING]
+====
+Avoid using only `java` functions or you can use the `java` functions to call the external services. Also, you can use the xref:service-orchestration/orchestration-of-opnapi-based-services.adoc[services orchestration features].
```
Comment (ricardozanini) [inline serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:170]: @hmanwani-rh I think this change lost the meaning of the warning. What we mean here is that users must avoid using `java` functions to call external services. They should use orchestration instead.

Comment (hmanwani-rh) [reply serverlessworkflow/modules/ROOT/pages/core/custom-functions-support.adoc:170]: @ricardozanini oh, thanks for letting me know. 
I've changed the warning as follows:

`Avoid using `java` functions to call the external services, however, you can use the xref:service-orchestration/orchestration-of-opnapi-based-services.adoc[services orchestration features].`
